File: packages/pdp-extractor-agent/src/llm/schemas.ts
/**
 * JSON schemas shared by provider adapters to enforce structured model output.
 *
 * OpenAI Responses (`text.format`) and Azure/AIStudio chat completions
 * (`response_format`) use the strict draft schemas. Gemini uses the OpenAPI
 * subset variants because `generationConfig.responseSchema` does not accept
 * `additionalProperties`.
 */

const KEYWORD_CATEGORY_ENUM = [
  "benefit",
  "effect",
  "ingredient",
  "usage",
  "faq",
  "review",
  "product",
  "price",
  "metric",
  "unknown"
] as const;

const OCR_LAYOUT_LINE_ROLE_ENUM = ["title", "body", "label", "value", "footnote"] as const;

/**
 * 레이아웃 관계 스키마. 중첩 대신 `parentId`로 평면을 유지하고, 선택 필드는
 * nullable 합집합으로 표현한다 — strict 모드는 모든 속성을 `required`에 요구하므로
 * "없음"을 표현할 다른 방법이 없다.
 */
const ocrLayoutGroupsSchema = {
  type: "array",
  description: "Layout relations visible in the image: groups of lines. No template is assumed.",
  items: {
    type: "object",
    additionalProperties: false,
    properties: {
      id: {
        type: "string",
        description: "Short id unique within this image, e.g. g1."
      },
      parentId: {
        type: ["string", "null"],
        description: "Id of the enclosing group, or null at top level."
      },
      title: {
        type: ["string", "null"],
        description: "The group's own heading text when the layout sets one apart, or null."
      },
      ordinal: {
        type: ["integer", "null"],
        description: "Number the layout actually printed for this group, or null. Never number groups yourself."
      },
      annotates: {
        type: ["string", "null"],
        description: "For a footnote, disclaimer, or test-condition group, the id of the group it qualifies. Otherwise null."
      },
      lines: {
        type: "array",
        items: {
          type: "object",
          additionalProperties: false,
          properties: {
            text: { type: "string" },
            role: {
              type: "string",
              enum: [...OCR_LAYOUT_LINE_ROLE_ENUM],
              description: "Layout function of the line, not its meaning: title, body, label (axis tick, legend name, caption, package spec), value (a measured number or badge), footnote."
            },
            pairedLabel: {
              type: ["string", "null"],
              description: "For a value line, the label it is printed against (its bar's tick, its badge caption). null when the layout does not pair it."
            }
          },
          required: ["text", "role", "pairedLabel"]
        }
      }
    },
    required: ["id", "parentId", "title", "ordinal", "annotates", "lines"]
  }
} as const;

/** Strict schema for vision OCR transcription responses. */
export const imageOcrJsonSchema = {
  type: "object",
  additionalProperties: false,
  properties: {
    images: {
      type: "array",
      items: {
        type: "object",
        additionalProperties: false,
        properties: {
          index: {
            type: "integer",
            description: "1-based number of the image as labeled in the prompt."
          },
          imageUrl: {
            type: "string",
            description: "Image URL exactly as labeled in the prompt."
          },
          text: {
            type: "string",
            description: "Faithful transcription of all visible text in reading order. Empty string when no readable product text."
          },
          confidence: {
            type: "number",
            description: "0-1 legibility/completeness confidence for this transcription."
          },
          groups: ocrLayoutGroupsSchema
        },
        required: ["index", "imageUrl", "text", "confidence", "groups"]
      }
    }
  },
  required: ["images"]
} as const;

/** Strict schema for OCR keyword/sentence classification responses. */
export const keywordClassificationJsonSchema = {
  type: "object",
  additionalProperties: false,
  properties: {
    keywords: {
      type: "array",
      items: {
        type: "object",
        additionalProperties: false,
        properties: {
          keyword: { type: "string" },
          category: { type: "string", enum: [...KEYWORD_CATEGORY_ENUM] },
          confidence: { type: "number" }
        },
        required: ["keyword", "category", "confidence"]
      }
    },
    sentenceInsights: {
      type: "array",
      items: {
        type: "object",
        additionalProperties: false,
        properties: {
          text: { type: "string" },
          category: { type: "string", enum: [...KEYWORD_CATEGORY_ENUM] },
          keywords: { type: "array", items: { type: "string" } },
          confidence: { type: "number" },
          evidenceIndex: {
            type: "integer",
            description: "1-based Evidence number in this request that this insight was derived from. 0 when unknown."
          }
        },
        required: ["text", "category", "keywords", "confidence", "evidenceIndex"]
      }
    },
    semanticFacts: {
      type: "object",
      additionalProperties: false,
      properties: {
        ingredients: { type: "array", items: { type: "string" } },
        benefits: { type: "array", items: { type: "string" } },
        effects: { type: "array", items: { type: "string" } },
        skinTypes: { type: "array", items: { type: "string" } },
        usageSteps: { type: "array", items: { type: "string" } },
        metricClaims: {
          type: "array",
          items: {
            type: "object",
            additionalProperties: false,
            properties: {
              label: { type: "string" },
              subject: { type: "string" },
              value: { type: "string" },
              unit: { type: "string" },
              timing: { type: "string" },
              period: { type: "string" },
              sample: { type: "string" },
              method: { type: "string" },
              caveat: { type: "string" },
              sentence: { type: "string" },
              sourceText: { type: "string" },
              evidenceIndex: {
                type: "integer",
                description: "1-based Evidence number in this request that this claim was derived from. 0 when unknown."
              }
            },
            required: ["label", "subject", "value", "unit", "timing", "period", "sample", "method", "caveat", "sentence", "sourceText", "evidenceIndex"]
          }
        },
        evidenceSentences: { type: "array", items: { type: "string" } },
        ingredientBenefitLinks: {
          type: "array",
          items: {
            type: "object",
            additionalProperties: false,
            properties: {
              ingredient: { type: "string" },
              benefit: { type: "string" },
              effect: { type: "string" },
              sentence: { type: "string" },
              sourceText: { type: "string" },
              evidenceIndex: {
                type: "integer",
                description: "1-based Evidence number in this request that this link was derived from. 0 when unknown."
              }
            },
            required: ["ingredient", "benefit", "effect", "sentence", "sourceText", "evidenceIndex"]
          }
        },
        citations: {
          type: "array",
          items: {
            type: "object",
            additionalProperties: false,
            properties: {
              type: { type: "string", enum: ["research", "article"] },
              title: { type: "string" },
              publisher: { type: "string" },
              author: { type: "string" },
              publishedAt: { type: "string" },
              url: { type: "string" },
              finding: { type: "string" },
              sourceText: { type: "string" },
              evidenceIndex: {
                type: "integer",
                description: "1-based Evidence number in this request that this citation was derived from. 0 when unknown."
              }
            },
            required: ["type", "title", "publisher", "author", "publishedAt", "url", "finding", "sourceText", "evidenceIndex"]
          }
        }
      },
      required: ["ingredients", "benefits", "effects", "skinTypes", "usageSteps", "metricClaims", "evidenceSentences", "ingredientBenefitLinks", "citations"]
    },
    summary: { type: "string" }
  },
  required: ["keywords", "sentenceInsights", "semanticFacts", "summary"]
} as const;

/** OpenAI Responses `text.format` payload for image OCR. */
export const openAiImageOcrTextFormat = {
  format: {
    type: "json_schema",
    name: "pdp_image_ocr",
    strict: true,
    schema: imageOcrJsonSchema
  }
} as const;

/** OpenAI Responses `text.format` payload for keyword classification. */
export const openAiKeywordClassificationTextFormat = {
  format: {
    type: "json_schema",
    name: "pdp_keyword_classification",
    strict: true,
    schema: keywordClassificationJsonSchema
  }
} as const;

/** Chat-completions `response_format` payload for image OCR. */
export const chatCompletionsImageOcrResponseFormat = {
  type: "json_schema",
  json_schema: {
    name: "pdp_image_ocr",
    strict: true,
    schema: imageOcrJsonSchema
  }
} as const;

/** Chat-completions `response_format` payload for keyword classification. */
export const chatCompletionsKeywordClassificationResponseFormat = {
  type: "json_schema",
  json_schema: {
    name: "pdp_keyword_classification",
    strict: true,
    schema: keywordClassificationJsonSchema
  }
} as const;

type GeminiSchema = Record<string, unknown>;

/** Converts a strict draft schema into the Gemini responseSchema OpenAPI subset. */
function toGeminiSchema(schema: unknown): GeminiSchema {
  if (Array.isArray(schema)) {
    return schema.map(toGeminiSchema) as unknown as GeminiSchema;
  }
  if (typeof schema !== "object" || schema === null) {
    return schema as GeminiSchema;
  }

  const source = schema as Record<string, unknown>;
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(source)) {
    if (key === "additionalProperties" || key === "description") {
      continue;
    }
    if (key === "type" && typeof value === "string") {
      result.type = value.toUpperCase();
      continue;
    }
    // Gemini OpenAPI 서브셋은 type 배열을 받지 않는다. strict 스키마가 "없음"을
    // 표현하려고 쓴 nullable 합집합을 이 방언의 nullable 표기로 옮긴다.
    if (key === "type" && Array.isArray(value)) {
      const concrete = value.find((item): item is string => typeof item === "string" && item !== "null");
      if (concrete) {
        result.type = concrete.toUpperCase();
      }
      if (value.includes("null")) {
        result.nullable = true;
      }
      continue;
    }
    if (key === "properties" && typeof value === "object" && value !== null) {
      result.properties = Object.fromEntries(
        Object.entries(value as Record<string, unknown>).map(([name, child]) => [name, toGeminiSchema(child)])
      );
      continue;
    }
    if (key === "items") {
      result.items = toGeminiSchema(value);
      continue;
    }
    result[key] = value;
  }

  return result;
}

/** Gemini responseSchema for image OCR. */
export const geminiImageOcrResponseSchema = toGeminiSchema(imageOcrJsonSchema);

/** Gemini responseSchema for keyword classification. */
export const geminiKeywordClassificationResponseSchema = toGeminiSchema(keywordClassificationJsonSchema);

File: packages/pdp-extractor-agent/src/pdp_extractor_agent/schemas.py
"""Provider-neutral strict JSON schemas and provider dialect projections.

The dictionaries deliberately remain JSON-shaped rather than becoming Pydantic
models: OpenAI/Azure and Gemini each require a different schema dialect on the
wire, and consumers need to inspect the frozen payloads directly.
"""

from __future__ import annotations

from typing import Any

from ._json_types import as_list, as_mapping
from .ocr.layout import IMAGE_OCR_JSON_SCHEMA as _image_ocr_json_schema

KEYWORD_CATEGORY_ENUM = [
    "benefit",
    "effect",
    "ingredient",
    "usage",
    "faq",
    "review",
    "product",
    "price",
    "metric",
    "unknown",
]

# The strict image schema lives with the OCR parser so both provider payload
# parsing and schema construction share the same five layout roles.
IMAGE_OCR_JSON_SCHEMA = _image_ocr_json_schema


def _array(items: dict[str, Any]) -> dict[str, Any]:
    return {"type": "array", "items": items}


def _strict_object(properties: dict[str, Any]) -> dict[str, Any]:
    return {
        "type": "object",
        "additionalProperties": False,
        "properties": properties,
        "required": list(properties),
    }


_KEYWORD = _strict_object(
    {
        "keyword": {"type": "string"},
        "category": {"type": "string", "enum": KEYWORD_CATEGORY_ENUM},
        "confidence": {"type": "number"},
    }
)
_SENTENCE_INSIGHT = _strict_object(
    {
        "text": {"type": "string"},
        "category": {"type": "string", "enum": KEYWORD_CATEGORY_ENUM},
        "keywords": _array({"type": "string"}),
        "confidence": {"type": "number"},
        "evidenceIndex": {
            "type": "integer",
            "description": "1-based Evidence number in this request that this insight was derived from. 0 when unknown.",
        },
    }
)
_METRIC_CLAIM = _strict_object(
    {
        "label": {"type": "string"},
        "subject": {"type": "string"},
        "value": {"type": "string"},
        "unit": {"type": "string"},
        "timing": {"type": "string"},
        "period": {"type": "string"},
        "sample": {"type": "string"},
        "method": {"type": "string"},
        "caveat": {"type": "string"},
        "sentence": {"type": "string"},
        "sourceText": {"type": "string"},
        "evidenceIndex": {
            "type": "integer",
            "description": "1-based Evidence number in this request that this claim was derived from. 0 when unknown.",
        },
    }
)
_INGREDIENT_BENEFIT_LINK = _strict_object(
    {
        "ingredient": {"type": "string"},
        "benefit": {"type": "string"},
        "effect": {"type": "string"},
        "sentence": {"type": "string"},
        "sourceText": {"type": "string"},
        "evidenceIndex": {
            "type": "integer",
            "description": "1-based Evidence number in this request that this link was derived from. 0 when unknown.",
        },
    }
)
_CITATION = _strict_object(
    {
        "type": {"type": "string", "enum": ["research", "article"]},
        "title": {"type": "string"},
        "publisher": {"type": "string"},
        "author": {"type": "string"},
        "publishedAt": {"type": "string"},
        "url": {"type": "string"},
        "finding": {"type": "string"},
        "sourceText": {"type": "string"},
        "evidenceIndex": {
            "type": "integer",
            "description": "1-based Evidence number in this request that this citation was derived from. 0 when unknown.",
        },
    }
)

KEYWORD_CLASSIFICATION_JSON_SCHEMA: dict[str, Any] = _strict_object(
    {
        "keywords": _array(_KEYWORD),
        "sentenceInsights": _array(_SENTENCE_INSIGHT),
        "semanticFacts": _strict_object(
            {
                "ingredients": _array({"type": "string"}),
                "benefits": _array({"type": "string"}),
                "effects": _array({"type": "string"}),
                "skinTypes": _array({"type": "string"}),
                "usageSteps": _array({"type": "string"}),
                "safetyTests": _array({"type": "string"}),
                "metricClaims": _array(_METRIC_CLAIM),
                "evidenceSentences": _array({"type": "string"}),
                "ingredientBenefitLinks": _array(_INGREDIENT_BENEFIT_LINK),
                "citations": _array(_CITATION),
            }
        ),
        "summary": {"type": "string"},
    }
)


def to_gemini_schema(schema: object) -> Any:
    """Project strict JSON Schema into Gemini's OpenAPI-like subset."""

    items = as_list(schema)
    if items is not None:
        return [to_gemini_schema(item) for item in items]
    mapping = as_mapping(schema)
    if mapping is None:
        return schema
    result: dict[str, Any] = {}
    for key, value in mapping.items():
        if key in {"additionalProperties", "description"}:
            continue
        if key == "type" and isinstance(value, str):
            result[key] = value.upper()
            continue
        value_items = as_list(value)
        if key == "type" and value_items is not None:
            concrete = next((item for item in value_items if isinstance(item, str) and item != "null"), None)
            if concrete:
                result[key] = concrete.upper()
            if "null" in value_items:
                result["nullable"] = True
            continue
        properties = as_mapping(value)
        if key == "properties" and properties is not None:
            result[key] = {name: to_gemini_schema(child) for name, child in properties.items()}
            continue
        if key == "items":
            result[key] = to_gemini_schema(value)
            continue
        result[key] = to_gemini_schema(value)
    return result


GEMINI_IMAGE_OCR_RESPONSE_SCHEMA: dict[str, Any] = to_gemini_schema(IMAGE_OCR_JSON_SCHEMA)
GEMINI_KEYWORD_CLASSIFICATION_RESPONSE_SCHEMA: dict[str, Any] = to_gemini_schema(KEYWORD_CLASSIFICATION_JSON_SCHEMA)

OPENAI_IMAGE_OCR_TEXT_FORMAT = {
    "format": {"type": "json_schema", "name": "pdp_image_ocr", "strict": True, "schema": IMAGE_OCR_JSON_SCHEMA}
}
OPENAI_KEYWORD_CLASSIFICATION_TEXT_FORMAT = {
    "format": {
        "type": "json_schema",
        "name": "pdp_keyword_classification",
        "strict": True,
        "schema": KEYWORD_CLASSIFICATION_JSON_SCHEMA,
    }
}
CHAT_COMPLETIONS_IMAGE_OCR_RESPONSE_FORMAT = {
    "type": "json_schema",
    "json_schema": {"name": "pdp_image_ocr", "strict": True, "schema": IMAGE_OCR_JSON_SCHEMA},
}
CHAT_COMPLETIONS_KEYWORD_CLASSIFICATION_RESPONSE_FORMAT = {
    "type": "json_schema",
    "json_schema": {"name": "pdp_keyword_classification", "strict": True, "schema": KEYWORD_CLASSIFICATION_JSON_SCHEMA},
}

# Existing JavaScript consumers use these exact spellings. Keeping aliases is
# cheap and avoids forcing Tasks 5/6 to translate shared configuration names.
imageOcrJsonSchema = IMAGE_OCR_JSON_SCHEMA
keywordClassificationJsonSchema = KEYWORD_CLASSIFICATION_JSON_SCHEMA
geminiImageOcrResponseSchema = GEMINI_IMAGE_OCR_RESPONSE_SCHEMA
geminiKeywordClassificationResponseSchema = GEMINI_KEYWORD_CLASSIFICATION_RESPONSE_SCHEMA
openAiImageOcrTextFormat = OPENAI_IMAGE_OCR_TEXT_FORMAT
openAiKeywordClassificationTextFormat = OPENAI_KEYWORD_CLASSIFICATION_TEXT_FORMAT
chatCompletionsImageOcrResponseFormat = CHAT_COMPLETIONS_IMAGE_OCR_RESPONSE_FORMAT
chatCompletionsKeywordClassificationResponseFormat = CHAT_COMPLETIONS_KEYWORD_CLASSIFICATION_RESPONSE_FORMAT
toGeminiSchema = to_gemini_schema
